src/Microsoft.SqlTools.ServiceLayer/ObjectManagement/ObjectTypes/Database/DatabaseHandler.cs

```diff
@@ -210,39 +219,141 @@ public override Task Save(DatabaseViewContext context, DatabaseInfo obj)
+                    }
+                    catch (SmoException ex)
+                    {
+                        // Revert to previous level of user access if we changed it before encountering the exception
+                        if (originalAccess != smoDatabase.DatabaseOptions.UserAccess)
+                        {
+                            smoDatabase.DatabaseOptions.UserAccess = originalAccess;
+                            smoDatabase.Alter(TerminationClause.RollbackTransactionsImmediately);
+                        }
+                        if (!(ex.InnerException is MissingObjectException) || (ex.InnerException is MissingObjectException && throwIfNotExist))
+                        {
+                            throw;
+                        }
+                    }
+                }
+                else
+                {
+                    throw new InvalidOperationException($"Provided URN '{objectUrn}' did not correspond to an existing database.");
+                }
+            }
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        /// Used to produce a TSQL script for detaching the specified database from a server.
+        /// </summary>
+        /// <param name="connectionUri">URI of the underlying connection for the detach request</param>
+        /// <param name="objectUrn">URN of the database to detach</param>
+        /// <param name="dropConnections">Whether to drop active connections to the database before detaching it</param>
+        /// <param name="updateStatistics">Whether to update the query optimization statistics related to the database</param>
+        /// <param name="throwIfNotExist">Whether to throw an exception if the specified database doesn't exist</param>
+        /// <returns>A string representing the generated TSQL script</returns>
+        public string ScriptDetach(string connectionUri, string objectUrn, bool dropConnections, bool updateStatistics, bool throwIfNotExist)
+        {
+            var builder = new StringBuilder();
+            ConnectionInfo connectionInfo = this.GetConnectionInfo(connectionUri);
+            using (var dataContainer = CreateDatabaseDataContainer(connectionUri, objectUrn))
+            {
+                try
+                {
+                    var smoDatabase = dataContainer.SqlDialogSubject as Database;
+                    if (smoDatabase != null)
+                    {
+                        builder.AppendLine("USE [master]");
+                        builder.AppendLine("GO");
+                        if (dropConnections)
+                        {
+                            builder.AppendLine($"ALTER DATABASE [{smoDatabase.Name}] SET SINGLE_USER WITH ROLLBACK IMMEDIATE");
+                            builder.AppendLine("GO");
+                        }
+                        builder.Append($"EXEC master.dbo.sp_detach_db @dbname = N'{smoDatabase.Name}'");
```

Comment:
Can't you just set the execution mode on SMO to script and then do it through there? We should avoid writing our own scripts (especially if they may differ from SMO) whenever possible

Reply:
Where can I find the docs for that? I was only seeing scripting options for Create statements

Reply:
Never mind, found the capture mode docs

Reply:
Yup. It's what we do for scripting in general.

https://learn.microsoft.com/en-us/sql/relational-databases/server-management-objects-smo/create-program/using-capture-mode?view=sql-server-ver16 for others' reference

Reply:
I tried the SMO scripting method, but apparently there's a bug with Detach in particular where it marks the database as dropped in the SMO code in scripting mode, even though the database still exists, which causes failures when trying to use any methods on the Database object afterwards. I'll stick with the manual script creation until that bug eventually gets fixed.